### ISSUE TITLE
Zarf-Seed-Registry Installation Fails on Init with Deployment is not ready: zarf/zarf-docker-registry error

### ISSUE THREAD

**Issue by erikschlegel** [defenseunicorns/zarf#592]
### Environment
Device and OS: Azure AKS Linux Ubuntu 20.04
App version: 0.19.6
Kubernetes distro being used: AKS Kubernetes V 1.22.6
Other:

### Steps to reproduce
1. Create an AKS Cluster
2. Run `zarf init --components git-server`.

### Expected result
Command succeeds and Zarf is initialized in the cluster.

### Actual Result
```
The following message repeats until the init run timesout
Deployment is not ready: zarf/zarf-docker-registry. 0 out of 1 expected pods are ready
```

output of `kubectl -n zarf get events`

```
LAST SEEN   TYPE      REASON              OBJECT                                       MESSAGE
15m         Normal    Scheduled           pod/injector                                 Successfully assigned zarf/injector to aks-agentpool-40722291-vmss000001
15m         Warning   FailedMount         pod/injector                                 MountVolume.SetUp failed for volume "zarf-payload-018" : failed to sync configmap cache: timed out waiting for the condition
15m         Warning   FailedMount         pod/injector                                 MountVolume.SetUp failed for volume "zarf-payload-023" : failed to sync configmap cache: timed out waiting for the condition
15m         Warning   FailedMount         pod/injector                                 MountVolume.SetUp failed for volume "zarf-payload-013" : failed to sync configmap cache: timed out waiting for the condition
15m         Warning   FailedMount         pod/injector                                 MountVolume.SetUp failed for volume "zarf-payload-008" : failed to sync configmap cache: timed out waiting for the condition
15m         Warning   FailedMount         pod/injector                                 MountVolume.SetUp failed for volume "zarf-payload-009" : failed to sync configmap cache: timed out waiting for the condition
15m         Warning   FailedMount         pod/injector                                 MountVolume.SetUp failed for volume "zarf-payload-019" : failed to sync configmap cache: timed out waiting for the condition
15m         Warning   FailedMount         pod/injector                                 MountVolume.SetUp failed for volume "stage1" : failed to sync configmap cache: timed out waiting for the condition
15m         Warning   FailedMount         pod/injector                                 MountVolume.SetUp failed for volume "zarf-payload-027" : failed to sync configmap cache: timed out waiting for the condition
15m         Warning   FailedMount         pod/injector                                 MountVolume.SetUp failed for volume "zarf-payload-007" : failed to sync configmap cache: timed out waiting for the condition
15m         Warning   FailedMount         pod/injector                                 (combined from similar events): MountVolume.SetUp failed for volume "zarf-payload-027" : failed to sync configmap cache: timed out waiting for the condition
15m         Normal    Scheduled           pod/zarf-docker-registry-789d8ddfb8-4pfgj    Successfully assigned zarf/zarf-docker-registry-789d8ddfb8-4pfgj to aks-agentpool-40722291-vmss000000
14m         Normal    Pulling             pod/zarf-docker-registry-789d8ddfb8-4pfgj    Pulling image "127.0.0.1:32178/library/registry:2.7.1"
14m         Warning   Failed              pod/zarf-docker-registry-789d8ddfb8-4pfgj    Failed to pull image "127.0.0.1:32178/library/registry:2.7.1": rpc error: code = Unknown desc = failed to pull and unpack image "127.0.0.1:32178/library/registry:2.7.1": failed to resolve reference "127.0.0.1:32178/library/registry:2.7.1": failed to do request: Head "https://127.0.0.1:32178/v2/library/registry/manifests/2.7.1": http: server gave HTTP response to HTTPS client
14m         Warning   Failed              pod/zarf-docker-registry-789d8ddfb8-4pfgj    Error: ErrImagePull
34s         Normal    BackOff             pod/zarf-docker-registry-789d8ddfb8-4pfgj    Back-off pulling image "127.0.0.1:32178/library/registry:2.7.1"
13m         Warning   Failed              pod/zarf-docker-registry-789d8ddfb8-4pfgj    Error: ImagePullBackOff
15m         Normal    SuccessfulCreate    replicaset/zarf-docker-registry-789d8ddfb8   Created pod: zarf-docker-registry-789d8ddfb8-4pfgj
15m         Normal    ScalingReplicaSet   deployment/zarf-docker-registry              Scaled up replica set zarf-docker-registry-789d8ddfb8 to 1
```

output of `kubectl -n zarf get all`
```
NAME                                        READY   STATUS             RESTARTS   AGE
pod/injector                                1/1     Running            0          40m
pod/zarf-docker-registry-789d8ddfb8-4pfgj   0/1     ImagePullBackOff   0          40m

NAME                           TYPE       CLUSTER-IP     EXTERNAL-IP   PORT(S)          AGE
service/zarf-docker-registry   NodePort   10.0.16.59     <none>        5000:31999/TCP   40m
service/zarf-injector          NodePort   10.0.144.122   <none>        5000:32178/TCP   40m

NAME                                   READY   UP-TO-DATE   AVAILABLE   AGE
deployment.apps/zarf-docker-registry   0/1     1            0           40m

NAME                                              DESIRED   CURRENT   READY   AGE
replicaset.apps/zarf-docker-registry-789d8ddfb8   1         1         0       40m
```

### Severity/Priority
😕  Blocked on deploying zarf packages to Azure AKS

**Comment by jeff-mccoy** [comment]
We haven't run into this issue on AKS before, but it looks like the node container runtime is trying to call localhost via https vs http, which is the standard for containerd and crio.  Is there any special config or other details about this provisioning that might change the container runtime by chance?

Would be helpful to run 'zarf destroy --confirm --remove-components' and then 'zarf init -l=trace'.

Sorry if markdown is weird, using the GitHub app right now.

**Comment by jeff-mccoy** [comment]
Thanks @erikschlegel this definitely looks like a CRI change on AKS we'll need to play with a bit, I'll spin up AKS again this weekend to see if we can reproduce it.  Are you provisioning AKS with IaC or via the Azure web interface?

**Comment by erikschlegel** [comment]
I'm provisioning the cluster directly through the Azure Portal. I confirmed that I was able to initialize zarf usong K8 version 1.21. I suspect this is a containerd issue as it's configured slightly different on AKS version 1.22+. This is worth checking out https://github.com/Azure/AgentBaker/pull/1369/files#diff-fb591749d1d0d4f2901cce9539d1ddae91a0389dfccd96bc7822ea0a6f9e35f2

**Comment by JasonvanBrackel** [comment]
@jeff-mccoy Any update on this?

**Comment by erikschlegel** [comment]
Hi @jeff-mccoy - AKS K8 version 1.21 is no longer available for deployment in the portal and now none of the AKS supported versions appear to work with zarf. Do you happen to have an update?

**Comment by dsmithbauer** [comment]
I can confirm this is still an issue.

> Hello, I am experiencing the same issue. Here is my output from kubectl describe pods:
> 
> ```
> Events:
>   Type     Reason     Age                From               Message
>   ----     ------     ----               ----               -------
>   Normal   Scheduled  48s                default-scheduler  Successfully assigned zarf/zarf-docker-registry-796648965f-2fjw5 to aks-demo-97972695-vmss000000
>   Normal   BackOff    19s (x2 over 48s)  kubelet            Back-off pulling image "127.0.0.1:32380/library/registry:2.7.1"
>   Warning  Failed     19s (x2 over 48s)  kubelet            Error: ImagePullBackOff
>   Normal   Pulling    8s (x3 over 48s)   kubelet            Pulling image "127.0.0.1:32380/library/registry:2.7.1"
>   Warning  Failed     8s (x3 over 48s)   kubelet            Failed to pull image "127.0.0.1:32380/library/registry:2.7.1": rpc error: code = Unknown desc = failed to pull and unpack image "1
> 27.0.0.1:32380/library/registry:2.7.1": failed to resolve reference "127.0.0.1:32380/library/registry:2.7.1": failed to do request: Head "https://127.0.0.1:32380/v2/library/registry/manifest
> s/2.7.1": http: server gave HTTP response to HTTPS client
>   Warning  Failed     8s (x3 over 48s)   kubelet            Error: ErrImagePull
> ```
> 
> version 21.2 in an Azure US Government AKS cluster

**Comment by jeff-mccoy** [comment]
Yeah they must be doing something special, containerd upstream still serves localhost on http and even tests for it, digging into this more this week:  https://github.com/containerd/containerd/blob/main/pkg/cri/server/image_pull_test.go

**Comment by Racer159** [comment]
While we work on this, as a note on a potential work around, you can also use an external registry as described here: https://docs.zarf.dev/docs/user-guide/the-zarf-cli/cli-commands/zarf_init

Under the `Integrations` tab in the Azure console you can tie an Azure Container Registry to your cluster and then init it with something like this:
`zarf init --registry-push-password={PASSWORD} --registry-push-username={USERNAME} --registry-url={REGISTRY}.azurecr.io` 

(you can also specify a separate `--registry-pull-password` and `--registry-pull-username` and can load the username(s)/password(s) from a zarf-config.toml as described here: https://docs.zarf.dev/docs/user-guide/the-zarf-cli/cli-commands/zarf_prepare_generate-config#synopsis)

**Comment by jeff-mccoy** [comment]
Put some notes in a new issue after digging around a bit.  Will try to test an older version of AKS later on tonight.

**Comment by jeff-mccoy** [comment]
Also tested on AKS 1.22.11 and seeing the same results.

**Comment by jeff-mccoy** [comment]
Added some new notes at https://github.com/Azure/AKS/issues/3303#issuecomment-1306328925.  

Looks like a bug in containerd that was patched 2 weeks ago. In the interim, [Acorn](https://acorn.io/) ran into this issue too and did what we've been trying to avoid ([modify the containerd config](https://github.com/acorn-io/acorn/pull/689)).  Root cause is the change @erikschlegel identified to allow containerd registry config overrides in AKS actually highlighted the underlying containerd issue.

**Comment by jeff-mccoy** [comment]
@cheruvu1 if you have any other data you'd like to drop on this issue, please leave it here. Thanks!

**Comment by jsburckhardt** [comment]
hi folks, as a workaround (that includes patch by [iceberg](https://github.com/containerd/containerd/pull/7393/files#diff-d8aa6f31074c380c292a7b0e9e8eeb132e2ff8fb8ae59aa14f42e2663a59def1) you can update containerd in your cluster.  
Just apt update and upgrade the node. I do it through a DaemonSet (example):

```yaml
apiVersion: apps/v1
kind: DaemonSet
metadata:
  name: update-cluster
  labels:
    app: update-cluster
spec:
  selector:
    matchLabels:
      app: update-cluster
  template:
    metadata:
      labels:
        app: update-cluster
    spec:
      containers:
      - name: update-cluster
        image: alpine
        imagePullPolicy: IfNotPresent
        command:
          - nsenter
          - --target
          - "1"
          - --mount
          - --uts
          - --ipc
          - --net
          - --pid
          - --
          - sh
          - -c
          - |
            # apt update and upgrade 
            export DEBIAN_FRONTEND=noninteractive apt update && apt upgrade -y
            sleep infinity
        securityContext:
          privileged: true
      dnsPolicy: ClusterFirst
      hostPID: true
```

![image](https://user-images.githubusercontent.com/18494471/204708248-681c060d-cae5-4b43-9799-1b136924ca73.png)

I was able to initialize zarf:
![image](https://user-images.githubusercontent.com/18494471/204704082-409da3bd-e82e-45ac-882b-af4403c611fd.png)

I did also deploy the Big Bang into AKS, but has couple of "bumps"
1. if aks has more than one node the zarf init fails because of a pvc already being claim (zarf registry)
2. gatekeeper has the label `control-plane: controller-manager` in their NS... preventing the hook to change the image ... after removing it, everything went smoothly.
![image](https://user-images.githubusercontent.com/18494471/204704540-8ed74725-8fb3-44cf-a087-f0acc06beaab.png)

**Comment by brandtkeller** [comment]
Incurred this problem while attempting to initialize zarf on a Nutanix kubernetes cluster running `v1.22.9` of kubernetes and `1.6.6` of containerd.

**Comment by ntwkninja** [comment]
NAMESPACE     NAME                                                         READY   STATUS             RESTARTS   AGE
kube-system   aws-node-4k6m4                                               1/1     Running            0          111m
kube-system   aws-node-9s5gv                                               1/1     Running            0          111m
kube-system   aws-node-qxgrn                                               1/1     Running            0          111m
kube-system   aws-node-termination-handler-79dbd64446-svh7r                1/1     Running            0          17h
kube-system   cluster-autoscaler-aws-cluster-autoscaler-65b5fb5c6f-tlf4d   1/1     Running            0          17h
kube-system   cluster-proportional-autoscaler-coredns-6ccfb4d9b5-gvrnc     1/1     Running            0          17h
kube-system   coredns-b5bccb59b-cqlxx                                      1/1     Running            0          17h
kube-system   coredns-b5bccb59b-xwsq9                                      1/1     Running            0          17h
kube-system   ebs-csi-controller-6b7c4667c4-kk6s2                          6/6     Running            0          17h
kube-system   ebs-csi-controller-6b7c4667c4-xv8wx                          6/6     Running            0          17h
kube-system   ebs-csi-node-dthh8                                           3/3     Running            0          17h
kube-system   ebs-csi-node-h88hj                                           3/3     Running            0          17h
kube-system   ebs-csi-node-mtxgt                                           3/3     Running            0          17h
kube-system   kube-proxy-dhkmt                                             1/1     Running            0          17h
kube-system   kube-proxy-t4mvh                                             1/1     Running            0          17h
kube-system   kube-proxy-zgwnl                                             1/1     Running            0          17h
kube-system   metrics-server-7cd9d56884-ffhbw                              1/1     Running            0          17h
zarf          injector                                                     1/1     Running            0          15m
zarf          zarf-docker-registry-7c64687bfc-547xp                        0/1     ImagePullBackOff   0          14m
```
```
Events:
  Type     Reason                  Age                From                     Message
  ----     ------                  ----               ----                     -------
  Normal   Scheduled               65s                default-scheduler        Successfully assigned zarf/zarf-docker-registry-7c64687bfc-547xp to ip-10-1-10-64.us-gov-east-1.compute.internal
  Normal   SuccessfulAttachVolume  63s                attachdetach-controller  AttachVolume.Attach succeeded for volume "pvc-82baccc9-bcf0-4bb6-8c18-fead693b07cd"
  Normal   Pulling                 30s (x2 over 55s)  kubelet                  Pulling image "127.0.0.1:30778/library/registry:2.8.1"
  Warning  Failed                  20s (x2 over 45s)  kubelet                  Failed to pull image "127.0.0.1:30778/library/registry:2.8.1": rpc error: code = Unknown desc = failed to pull and unpack image "127.0.0.1:30778/library/registry:2.8.1": failed to resolve reference "127.0.0.1:30778/library/registry:2.8.1": failed to do request: Head "https://127.0.0.1:30778/v2/library/registry/manifests/2.8.1": net/http: TLS handshake timeout
  Warning  Failed                  20s (x2 over 45s)  kubelet                  Error: ErrImagePull
  Normal   BackOff                 5s (x2 over 44s)   kubelet                  Back-off pulling image "127.0.0.1:30778/library/registry:2.8.1"
  Warning  Failed                  5s (x2 over 44s)   kubelet                  Error: ImagePullBackOff
```

**Comment by ntwkninja** [comment]
EKS v1.23 works without issue because it is still using docker vs. containerd

**Comment by jsburckhardt** [comment]
quick update, containerd was updated and I could deploy bigbang 1.48.0. Ticket can be closed ;) thanks for your help
![image](https://user-images.githubusercontent.com/18494471/221087730-cc0b6108-6d58-49ad-92a9-c6facebda398.png)

**Comment by ntwkninja** [comment]
> Tracking EKS AMI containerd update: [awslabs/amazon-eks-ami#1162](https://github.com/awslabs/amazon-eks-ami/issues/1162)

Upstream issue has been closed and @brianrexrode has successfully tested zarf v0.25.2 with EKS V1.26

**Comment by Racer159** [comment]
@AbrohamLincoln thanks for the note! We're looking at exploring other options too, and for others this will affect newer versions of containerd 1.7 (>1.7.7) (and 2.0 if anyone is on the betas) as well.

**Comment by philip-h-iversen-boeing** [comment]
Seeing similar behavior on EKS `1.26`  - `1.29` all of which are using containerd `1.7.11`. Zarf init was previously (late January) working on EKS `1.26` with containerd `1.7.2`. Confirmed that going back to EKS `1.23` with docker runtime `20.10.25` does work, but that version is out of service.

**Comment by toby-darkwolf** [comment]
Seeing this on RKE2 1.27 as well with upgrade to containerd v1.7.7

**Comment by jnt2007** [comment]
Hi guys. Any plans here? Still having a problem with k8s 1.27+ versions with containerd `1.7.11`.

Any recommendations from the community on how we can "tune" a containerd config to avoid this issue?

**Comment by philiversen** [comment]
> Hi guys. Any plans here? Still having a problem with k8s 1.27+ versions with containerd `1.7.11`.
> 
> Any recommendations from the community on how we can "tune" a containerd config to avoid this issue?

Commenting out the containerd config lines mentioned in [this post](https://github.com/defenseunicorns/zarf/issues/592#issuecomment-1932652385) got things working again for me.

**Comment by mjnagel** [comment]
Also ran into this issue on newer RKE2 versions. It seems linked back to [this commit](https://www.github.com/k3s-io/k3s/commit/c45524e66237bdb491ebb0afc633a5bd6d3a76c8) which introduced the `config_path` line. The containerd update to 1.7.7+ actually did not immediately cause the issue (RKE2 1.29.0 works and is on containerd 1.7.11), it was specifically the introduction of that containerd config line 👀.

Just for reference affected versions of k3s/rke2 appear to be 1.29.1+, 1.28.6+, and 1.27.10+. Definitely curious if there is anything to address this on the zarf side or if this should make its way into the docs as a recommended pre-req/setup for the cluster?

**Comment by jeff-mccoy** [comment]
All things coming around, we may need to look to a way to avoid the localhost/http behavior since containerd has introduced bugs multiple times for this in the past year or so. https://github.com/containerd/containerd/pull/9188

**Comment by lucasrod16** [comment]
A new issue has been opened against containerd to address this: https://github.com/containerd/containerd/issues/10014

**Comment by JasonRodriguez1474** [comment]
> > Hi guys. Any plans here? Still having a problem with k8s 1.27+ versions with containerd `1.7.11`.
> > Any recommendations from the community on how we can "tune" a containerd config to avoid this issue?
> 
> Commenting out the containerd config lines mentioned in [this post](https://github.com/defenseunicorns/zarf/issues/592#issuecomment-1932652385) got things working again for me.

Does anyone know if this config line fix is in the default zarf init version of k3s? 
I've been playing around with this issue and am observing the ImagePullBackOff for the zarf-docker-registry in various versions of containerd when I connect to a cluster I started on my own, but not when I use the K3S that Zarf initiates as part of zarf init. 

| Kubernetes Version | Container-Runtime | CR version  | zarf-docker-registry init | Zarf K3S                  |
|--------------------|-------------------|-------------|---------------------------|---------------------------|
| v1.28.8+k3s1       | containerd        | 1.7.10      | Fails                     | No, Rancher Desktop       |
| v1.28.8+k3s1       | docker            | 24.0.7      | Works                     | No, Rancher Desktop       |
| v1.28.8+k3s1       | containerd        | 1.7.11-k3s2 | Fails                     | No, standard K3S install  |
| v1.28.4+k3s2       | containerd        | 1.7.7-k3s1  | Works                     | Yes                       |

Error message observed when it fails
```
err="failed to \"StartContainer\" for \"docker-registry\" with ImagePullBackOff: \"Back-off pulling image \\\"127.0.0.1:31585/library/registry:2.8.3\\\"\"" pod="zarf/zarf-docker-registry-6c9d98c95d-z7k6r" 
```

**Comment by lucasrod16** [comment]
A fix has been merged in and backported to `1.7` and pending `1.6` for containerd: https://github.com/containerd/containerd/pull/10109

I have not tested the fix myself yet, but hoping it resolves this issue 🤞

**Comment by lucasrod16** [comment]
Containerd fix has been released in `v1.7.16`: https://github.com/containerd/containerd/releases/tag/v1.7.16
